ROS Node Implementation for timestamping

4c3y opened the pull request:
## Features
- Implemented a C++ [sysfs](https://en.wikipedia.org/wiki/Sysfs) interface to access Raspberry Pi's Hardware PWM timers.
- Implemented TimestampManager which polls for new timestamps generated by the kernelmodule.
- Implemented ROS Node with timestamp publisher.
- Added unittests with sysfs mock
- Added docs in header files

4c3y committed:
Added .gitignore, Moved kernel module

4c3y committed:
Added catkin package, Added Timestamp.msg

4c3y committed:
Added project structure

4c3y committed:
Implemented SysfsPwm, Added SignalHandler, Refactoring

4c3y committed:
Implemented Node() partially

4c3y committed:
Added TimestampManager, First working node implementation

4c3y committed:
Bugfixed IsRunning()

4c3y committed:
Implemented Sysfs::IsExported()

4c3y committed:
Improved node init, Implemented Cleanup logic

4c3y committed:
Small code cleanup

4c3y committed:
Implemented SetFrequency()

4c3y committed:
Refactoring, Removed global variable, Added test entry point

4c3y committed:
Added gmock in CMakeLists.txt

4c3y committed:
Reworked SysfsPwm and IPwmSubsystem, Added test_mock_pwm_subsystem.cc

4c3y committed:
Created subsystem unittests

4c3y committed:
replaced info with warning if pwm subsystem is already running

4c3y committed:
Added Syspwm reset logic

4c3y committed:
Fixed Unittests, Moved hardcoded values to config, created launch file, Updated Node::init()

4c3y committed:
Moved hardcoded sysfs path to config

4c3y committed:
Fixed Node::Init() in unittests, Added Unittest with force reset enabled

4c3y committed:
Formatted all files

4c3y committed:
Added docs in header files + refactoring

4c3y committed:
Added empty dependencies.rosinstall

4c3y committed:
Moved dependencies.rosinstall, Added CI dependencies

michaelpantic commented on time_stamper_ros/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+cmake_minimum_required(VERSION 3.0.2)
```
We normally use catkin simple - please switch to it

michaelpantic commented on time_stamper_ros/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+        src/Main.cpp
+        )
+
+#find_library(WIRINGPI_LIBRARIES NAMES wiringPi)
```
Remove unused stuff

michaelpantic commented on time_stamper_ros/include/IPwmSubsystem.h:
```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+#include <fcntl.h>
+class IPwmSubsystem {
```
Interfaces are a good place to document interfaces :-)

Please add comments and documentation for the interfacce and the individual functions.
Naming of the functions is pretty good, but not always self-evident. E.g. from only code its not clear what the difference between SetDutyCycle and SetDutyCycleRaw is, or what the write function is for

michaelpantic replied:
ah, I just saw all the (good) comments in the implemented class. Maybe its better to move them to the interface?

michaelpantic commented on time_stamper_ros/include/IPwmSubsystem.h:
```diff
@@ -0,0 +1,20 @@
+  virtual bool IsRunning() = 0;
+  virtual bool Start() = 0;
+  virtual bool Stop() = 0;
+  virtual bool SetFrequency(int hz) = 0;
```
use uint32 or 16 (what are the maximums?)

btw, is frequency always integer?

michaelpantic commented on time_stamper_ros/msg/Timestamp.msg:
```diff
@@ -0,0 +1 @@
+float64 timestamp
```
Ros useually uses ros::Time for timestamps. Any reason not to add it too?

michaelpantic commented on time_stamper_ros/src/Node.cpp:
```diff
@@ -0,0 +1,71 @@
+  if (forceReset) {
+    pwm_subsystem_.Reset();
+    ROS_INFO("Reset pwm");
+    usleep(1e3 * 10);
```
needed for lowlevel things? Add comment

michaelpantic commented on time_stamper_ros/src/SysfsPwm.cpp:
```diff
@@ -0,0 +1,120 @@
+bool SysfsPwm::Reset() {
+  bool isReset = Unexport()
+      && Export()
+      && Write("/pwm0/period", "10000000")
```
magic strings, should be class variables with comments

michaelpantic replied:
For putting together the file paths, you could use std::filesystem::path for the individual parts, eg. pwm0, period etc as static class variables, and then concatenate them together.

https://en.cppreference.com/w/cpp/filesystem/path

michaelpantic commented on time_stamper_ros/src/SysfsPwm.cpp:
```diff
@@ -0,0 +1,120 @@
+    return false;
+  }
+
+  if (a == 1) {
```
you can just return a==1 as last statement here

michaelpantic commented on time_stamper_ros/src/SysfsPwm.cpp:
```diff
@@ -0,0 +1,120 @@
+}
+
+bool SysfsPwm::ChangeDutyCycle(int percentage) {
+  std::string value_str = std::to_string(5000000);
```
static class member for magic strings

michaelpantic commented on time_stamper_ros/src/SysfsPwm.cpp:
```diff
@@ -0,0 +1,120 @@
+}
+
+bool SysfsPwm::SetFrequency(int hz) {
+  int freq = (int) 1e9 / hz;
```
hmm can this ever go wrong? Test input range to make sure it doesnt

michaelpantic replied:
(think about very large, negative or 0 hz

michaelpantic commented on time_stamper_ros/src/SysfsPwm.cpp:
```diff
@@ -0,0 +1,120 @@
+    return false;
+  }
+
+  DIR *pDir = opendir(path);
```
could use https://www.gnu.org/software/libc/manual/html_node/Testing-File-Type.html

michaelpantic commented on time_stamper_ros/test/test_mock_pwm_subsystem.cc:
```diff
@@ -0,0 +1,124 @@
+TEST(MockPwmSubsystem, TestFrequencyValueValidation) {
+  //Precondition: Subsystem is in any valid state
+  MockPwmSubsystem mock_pwm_subsystem;
+  //TODO define valid range
```
do todos :-)

michaelpantic commented on time_stamper_ros/test/test_mock_pwm_subsystem.cc:
```diff
@@ -0,0 +1,124 @@
+TEST(MockPwmSubsystem, TestNodeInitSanityCheck) {
+  //Precondition: Subsystem is not exported but running, which is undefined behavior and should stop node instantly
+
+  //TODO implement
```
same

michaelpantic reviewed (changes requested):
Nice and readable code, well done!

Of course I always find things to complain about, see comments.
Let me know when you addressed them, then I have a second look!

4c3y committed:
Extracted String constants, Added Documentation in IPwmSubsystem, Changed Documentation in SysPwm

4c3y committed:
Fixed unittests to work on ubuntu 18

4c3y committed:
Changed to C++17 to avoid inline warnings

4c3y committed:
Removed unused code in CMake, Moved timing function to SysPwm

4c3y committed:
Improved error messages, Fixed file descriptor leak, Bugfixed Node::Cleanup()

4c3y committed:
Bugfix in SysfsPwm::SetFrequency()

4c3y committed:
Extracted Filesystem interface, Made SysfsPwm testable with filesystem mock. Moved constants from SysfsPwm to IPwmSubsystem

4c3y committed:
Refactored Filesystem::DirectoryExists(), Implemented Sysfs::ChangeDutyCycle + Refactoring

4c3y committed:
Fixed typos

4c3y committed:
Added top level README.md

4c3y committed:
Converted CMakeLists.txt to catkin_simple format

4c3y commented:
Addressed all requested changes and added additional documentation. As we discussed last week I extracted a filesystem interface from the pwm subsystem interface to improve test coverage.

4c3y committed:
Refactoring

michaelpantic commented on time_stamper_ros/test/Tests/test_mock_filesystem.cc:
```diff
@@ -0,0 +1,55 @@
+
+  //Expectation: Subsystem should filter invalid values
+
+  bool setFrequencyResult = sysfs_pwm.SetFrequency(IPwmSubsystem::PWM_MAXSPEED_HZ + 1);
```
Unit tests should be as minimal as possible, so better to break up into smaller ones.

Otherwise good tests!

michaelpantic replied:
see e.g https://osherove.com/blog/2005/4/3/a-unit-test-should-test-only-one-thing.html

michaelpantic reviewed (approved):
Well done! Code looks good, the new unit tests are also good.

Just see my comment about breaking them up - its not wrong to have only one test statement per unit tests, so the CI results are easier to track. Once the unit tests are broken up, just merge!

4c3y committed:
Split Unittests
